bundles/config/org.eclipse.smarthome.config.xml/src/main/java/org/eclipse/smarthome/config/xml/osgi/XmlDocumentBundleTracker.java

```diff
@@ -69,174 +83,297 @@
+     * a fragment. So the fragment bundle resources can override the
+     * host bundles resources.
+     *
+     * @param xmlDocumentPaths
+     * @param bundle
+     * @return
+     */
+    private Collection<URL> filterPatches(Enumeration<URL> xmlDocumentPaths, Bundle bundle) {
+        List<URL> hostResources = new ArrayList<URL>();
+        List<URL> fragmentResources = new ArrayList<URL>();
+
+        while (xmlDocumentPaths.hasMoreElements()) {
+            URL path = xmlDocumentPaths.nextElement();
+            if (bundle.getEntry(path.getPath()) != null && bundle.getEntry(path.getPath()).equals(path)) {
+                hostResources.add(path);
+            } else {
+                fragmentResources.add(path);
+            }
+        }
+        if (!fragmentResources.isEmpty()) {
+            Map<String, URL> helper = new HashMap<String, URL>();
+            for (URL url : hostResources) {
+                helper.put(url.getPath(), url);
+            }
+            for (URL url : fragmentResources) {
+                helper.put(url.getPath(), url);
+            }
+            return helper.values();
+        }
+        return hostResources;
+    }
+
+    /**
+     * Checks for the existence of a given resource inside the bundle and its attached fragments.
+     *
+     * Helper method which can be used in {@link #isBundleRelevant(Bundle)}.
+     *
+     * @param bundle
+     * @param path the directory name to look for
+     * @return <code>true</code> if the bundle or one of its attached fragments contain the given directory
+     */
+    private final boolean isResourcePresent(Bundle bundle, String path) {
+        return bundle.getEntry(path) != null;
+    }
+
+    /**
+     * Add a bundle which potentially needs to be processed.
+     *
+     * This method should be called in order to queue a new bundle for asynchronous processing.
+     * It can be used e.g. by a {@link BundleTracker}, detecting a new bundle.
+     * <p>
+     * If the bundle actually will be put into the queue depends on the presence of the corresponding XML configuration
+     * directory.
+     *
+     * @param bundle
+     */
+    private void addingBundle(Bundle bundle) {
+        if (!isBundleRelevant(bundle)) {
+            finishBundle(bundle);
+            return;
+        }
+        ScheduledFuture<?> future = scheduler.schedule(() -> {
+            processBundle(bundle);
+        }, 0, TimeUnit.SECONDS);
+        queue.put(bundle, future);
+    }
+
+    private void finishBundle(Bundle bundle) {
+        queue.remove(bundle);
+        finishedBundles.add(bundle);
+        registerReadyMarker(bundle);
+
+        Set<Bundle> remainingBundles = getRemainingBundles();
+        if (!remainingBundles.isEmpty()) {
+            logger.trace("Remaining bundles with {}: {}", xmlDirectory, remainingBundles);
+        } else {
+            logger.debug("Finished loading bundles with {}", xmlDirectory);
```

Comment:
I seem to get one of these messages for every bundle in my installation...
Not really sure if we are at all expected to hit this line so often (if so, better reduce to trace) or if there is some issue in the processing and we should not run out of remainingBundles at startup continuously.